libcudacxx/include/cuda/std/__atomic/functions/cuda_ptx_derived.h

```diff
@@ -33,6 +31,183 @@ _LIBCUDACXX_BEGIN_NAMESPACE_STD
+{
+  uint32_t* __aligned     = (uint32_t*) ((intptr_t) __ptr & ~(sizeof(uint32_t) - 1));
+  const uint32_t __offset = uint32_t((intptr_t) __ptr & (sizeof(uint32_t) - 1)) * 8;
+
+  uint32_t __value = 0;
+
+  __cuda_atomic_load(__aligned, __value, _Order{}, __atomic_cuda_operand_b32{}, _Sco{}, __atomic_cuda_mmio_disable{});
+  __dst = static_cast<_Type>(__value >> __offset);
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE bool
+__cuda_atomic_compare_exchange(_Type* __ptr, _Type& __dst, _Type __cmp, _Type __op, _Order, _Operand, _Sco)
+{
+  uint32_t* __aligned           = (uint32_t*) ((intptr_t) __ptr & ~(sizeof(uint32_t) - 1));
+  constexpr uint32_t __sizemask = (1 << (sizeof(_Type) * 8)) - 1;
+  const uint32_t __offset       = uint32_t((intptr_t) __ptr & (sizeof(uint32_t) - 1)) * 8;
+  const uint32_t __valueMask    = __sizemask << __offset;
+  const uint32_t __windowMask   = ~__valueMask;
+
+  // Algorithm for 8b CAS with 32b intrinsics
+  // __old = __window[0:32] where [__cmp] resides within any of the potential offsets
+  // First CAS attempt 'guesses' that the masked portion of the window is 0x00.
+  uint32_t __old = (uint32_t(__cmp) << __offset);
+
+  // Reemit CAS instructions until either of two conditions are met
+  while (1)
+  {
+    // Combine the desired value and most recently fetched expected masked portion of the window
+    uint32_t __attempt = (__old & __windowMask) | (uint32_t(__op) << __offset);
+
+    if (__cuda_atomic_compare_exchange(
+          __aligned, __old, __old, __attempt, _Order{}, __atomic_cuda_operand_b32{}, _Sco{}))
+    {
+      // CAS was successful
+      return true;
+    }
+    auto __old_value = static_cast<_Type>((__old & __valueMask) >> __offset);
+    // The expected value no longer matches inside the CAS.
+    if (__old_value != __cmp)
+    {
+      __dst = __old_value;
+      break;
+    }
+  }
+  return false;
+}
+
+// Lower level fetch_update that bypasses memorder dispatch
+template <class _Type, class _Fn, class _Sco, class _Order, class _Operand>
+_CCCL_DEVICE _Type __cuda_atomic_fetch_update(_Type* __ptr, const _Fn& __op, _Order, _Operand, _Sco)
+{
+  _Type __expected = 0;
+  __atomic_load_cuda(__ptr, __expected, __ATOMIC_RELAXED, _Sco{});
+  _Type __desired = __op(__expected);
+  while (!__cuda_atomic_compare_exchange(__ptr, __expected, __expected, __desired, _Order{}, _Operand{}, _Sco{}))
+  {
+    __desired = __op(__expected);
+  }
+  return __expected;
+}
+
+template <class _Type, class _Order, class _Operand, class _Sco, __cuda_atomic_enable_non_native_bitwise<_Operand> = 0>
+static inline _CCCL_DEVICE void
+__cuda_atomic_store(_Type* __ptr, _Type __val, _Order, _Operand, _Sco, __atomic_cuda_mmio_disable)
+{
+  // Store requires cas on 8/16b types
+  __cuda_atomic_fetch_update(
+    __ptr,
+    [__val](_Type __old) {
+      return __val;
+    },
```

Comment:
We should have a proper function object with an overloaded call operator. Lambdas tend to carry a significant compile time overhead as they require a unique type to be instantiated

Reply:
Dropping this because we need to capture the value anyhow